src/features/edit/edit-label-ui.ts

```diff
@@ -0,0 +1,230 @@
+    @inject(TYPES.ViewerOptions) protected viewerOptions: ViewerOptions;
+    @inject(TYPES.DOMHelper) protected domHelper: DOMHelper;
+    @inject(TYPES.IEditLabelValidator) @optional() public labelValidator: IEditLabelValidator;
+
+    protected inputElement: HTMLInputElement;
+    protected label?: EditableLabel & SModelElement;
+    protected labelElement: HTMLElement | null;
+    protected validationTimeout?: number = undefined;
+    protected isActive: boolean = false;
+
+    protected get labelId() { return this.label ? this.label.id : 'unknown'; }
+
+    protected initializeContents(containerElement: HTMLElement): void {
+        containerElement.style.position = 'absolute';
+        this.inputElement = document.createElement('input');
+        this.inputElement.addEventListener('keydown', (event) => this.hideIfEscapeEvent(event));
+        this.inputElement.onblur = () => window.setTimeout(() => this.applyLabelEdit(), 200);
+        this.inputElement.addEventListener('keydown', (event) => this.applyLabelEditIfEnterEvent(event));
+        this.inputElement.onkeyup = (event) => this.performLabelValidation(event, this.inputElement.value);
+        containerElement.appendChild(this.inputElement);
+    }
+
+    protected hideIfEscapeEvent(event: KeyboardEvent): any {
+        if (matchesKeystroke(event, 'Escape')) { this.hide(); }
+    }
+
+    protected applyLabelEditIfEnterEvent(event: KeyboardEvent): any {
+        if (matchesKeystroke(event, 'Enter')) {
+            this.applyLabelEdit();
+        }
+    }
+
+    private applyLabelEdit() {
+        if (!this.isActive) {
+            return;
+        }
+        this.actionDispatcherProvider()
+            .then(actionDispatcher => actionDispatcher.dispatch(new ApplyLabelEditAction(this.labelId, this.inputElement.value)))
+            .catch((reason) => this.logger.error(this, 'No action dispatcher available to execute apply label edit action', reason));
+        this.hide();
+    }
+
+    protected performLabelValidation(event: KeyboardEvent, value: string) {
+        if (this.validationTimeout) {
+            window.clearTimeout(this.validationTimeout);
+        }
+        this.validationTimeout = window.setTimeout(() => this.validateLabel(value), 200);
+    }
+
+    protected validateLabel(value: string) {
+        if (this.labelValidator && this.label) {
+            this.labelValidator.validate(value, this.label)
+                .then(result => this.showValidationResult(result))
+                .catch((reason) => this.logger.error(this, 'Error validating edited label', reason));
+        }
+    }
+
+    protected showValidationResult(result: EditLabelValidationResult) {
+        this.clearValidationResult();
+        if (result.message) {
+            this.containerElement.setAttribute('data-balloon', result.message);
+            this.containerElement.setAttribute('data-balloon-pos', 'up-left');
+            this.containerElement.setAttribute('data-balloon-visible', 'true');
```

Comment:
Yes, sort of... they are related to [ballon.css](https://kazzkiq.github.io/balloon.css/), which we load in the class diagram example. Even though it is "just" some HTML attributes that we essentially add, you're probably right that we should extract that into application-specific code. Should we allow to inject a new type eg named `LabelEditValidationUi` which is invoked here and that is implemented in the application-specific code so that it can reuse existing UI frameworks?